DX: improve How zkApps Work content (#484)
o1-labs/docs2: docs/zkapps/how-zkapps-work.mdx

zkApps run in the latest versions of [Chrome](https://www.google.com/chrome/index.html), [Firefox](https://www.mozilla.org/en-US/firefox/new/), [Edge](https://microsoft.com/edge), and [Brave](https://brave.com/) web browsers.

zkApps are written in TypeScript using the Mina zkApp CLI. 

A zkApp consists of two parts: 

- A smart contract

- A user interface (UI) for users to interact with the zkApp

<figure>
  <img src="/img/3_zkApps_Structure.jpg" width="100%" alt="Diagram showing two parts of a zkApp" />
</figure>

- The term _smart contract_ refers to the code written with SnarkyJS.

- The term _zkApp_ refers to the UI + the smart contract.

## Zero knowledge-based smart contracts

zkApps are based on zero knowledge proofs (zk-SNARKs). As a zkApp developer, you use the zkApp CLI to scaffold and deploy your project. 

> A circuit is the method from which a prover function and a corresponding verifier function are derived during the build process.

Author: what is a circuit now? provable something?

Reviewer: It's now `Provable code`. You could rephrase this paragraph to something like this

> Provable code is written by using SnarkyJS, it generated a prover function and a corresponding verifier function that are derived during the build process.

> A circuit is the method from which a prover function and a corresponding verifier function are derived during the build process.

Author: ```suggestion
Provable code is written by using SnarkyJS and generates a prover function and a corresponding verifier function that are derived during the build process.
```

> The prover function is the function that executes a smart contract's custom logic and runs in a user's web browser as part of the zkApp. When interacting with a zkApp UI, users enter any data (for example, buy ABC for y price) that is required as input to the prover function, which then generates a zero knowledge proof.

Reviewer: probably worth mentioning that the prover function actually generated a proof of the executed code!

</figure>

Both private inputs and public inputs represent data that must be provided to the prover function when it runs in the user's web browser.

Private inputs are not required again. Because public inputs must also be provided to the verifier function when it runs on the Mina network, public inputs are not used for data that you want to remain private.

The verifier function validates whether a zero knowledge proof successfully passes all the constraints defined in the prover function. The verifier function _always_ runs quickly and efficiently, irrespective of the prover function's complexity.

Within the Mina network, Mina acts as the verifier and runs the verifier function. 

<figure>
  <img src="/img/5_zkApps_Verifier_Function.jpg" width="90%" alt="Diagram showing verifier function validation" />
</figure>

## Prover Function and Verification Key

After you write your smart contract, run the `npm run build` command. The build process compiles the TypeScript code into JavaScript and outputs the `smart_contract.js` file. 

From this file, you can:

- Run a prover function to run your smart contract
- Generate a verification key to deploy your smart contract

> While the prover function runs in a user's web browser, the verification key lives on-chain for a given zkApp account and is used by the Mina network to verify that a zero knowledge proof has met all constraints defined in the prover. A verification key is required to create a zkApp account.

Reviewer: You can also use the verifier function (or verification key) to verify proofs off-chain, but yes the main usage for the verifier function is the Mina network

> While the prover function runs in a user's web browser, the verification key lives on-chain for a given zkApp account and is used by the Mina network to verify that a zero knowledge proof has met all constraints defined in the prover. A verification key is required to create a zkApp account.

Author: ```suggestion
While the prover function runs in a user's web browser, the verification key lives on-chain for a given zkApp account and is used by the Mina network to verify that a zero knowledge proof has met all constraints defined in the prover. A verification key is required to create a zkApp account.

You can also use the verifier function or verification key to verify proofs off-chain.
```

## Deploy a smart contract

> Smart contracts are deployed to the Mina network using the Mina zkApp CLI. The deployment process sends a transaction that contains the verification key to an address on the Mina blockchain.

Reviewer: probably worth mentioning that the verification key is essentially the verifier function (or an artifact of it)

<figure>
  <img src="/img/6_zkApps_DeploySmartContract.jpg" width="75%" alt="Diagram showing deployment of a Mina zkApp smart contract" />
</figure>

When a Mina address contains a verification key, it acts as a zkApp account.

A regular Mina account can receive any transactions. 

> Because zkApps are Mina Protocol's smart contracts powered by zero knowledge proofs, a zkApp account can successfully receive only transactions that contain a proof that satisfies the verifier function. Any transactions that do not pass the verifier function are rejected by the Mina network.

Reviewer: I don't think this is 100% correct - you can specify custom permissions on smart contract accounts and by doing so, you can also send transactions to it that are not authorized by a proof

zkApps run in the latest versions of [Chrome](https://www.google.com/chrome/index.html), [Firefox](https://www.mozilla.org/en-US/firefox/new/), [Edge](https://microsoft.com/edge), and [Brave](https://brave.com/) web browsers.

zkApps are written in TypeScript using the Mina zkApp CLI. 

A zkApp consists of two parts: 

- A smart contract

- A user interface (UI) for users to interact with the zkApp

<figure>
  <img src="/img/3_zkApps_Structure.jpg" width="100%" alt="Diagram showing two parts of a zkApp" />
</figure>

- The term _smart contract_ refers to the code written with SnarkyJS.

- The term _zkApp_ refers to the UI + the smart contract.

## Zero knowledge-based smart contracts

zkApps are based on zero knowledge proofs (zk-SNARKs). As a zkApp developer, you use the zkApp CLI to scaffold and deploy your project. 

Provable code is written by using SnarkyJS and generates a prover function and a corresponding verifier function that are derived during the build process.

> The prover function is the function that executes a smart contract's custom logic and runs in a user's web browser as part of the zkApp. When interacting with a zkApp UI, users enter any data (for example, buy ABC for y price) that is required as input to the prover function, which then generates a zero knowledge proof.

Author: ```suggestion
The prover function is the function that executes a smart contract's custom logic and runs in a user's web browser as part of the zkApp. The prover function generates a proof of the executed code. 

When interacting with a zkApp UI, users enter any data (for example, buy ABC for y price) that is required as input to the prover function, which then generates a zero knowledge proof.
```

Within the Mina network, Mina acts as the verifier and runs the verifier function. 

<figure>
  <img src="/img/5_zkApps_Verifier_Function.jpg" width="90%" alt="Diagram showing verifier function validation" />
</figure>

## Prover Function and Verification Key

After you write your smart contract, run the `npm run build` command. The build process compiles the TypeScript code into JavaScript and outputs the `smart_contract.js` file. 

From this file, you can:

- Run a prover function to run your smart contract
- Generate a verification key to deploy your smart contract

While the prover function runs in a user's web browser, the verification key lives on-chain for a given zkApp account and is used by the Mina network to verify that a zero knowledge proof has met all constraints defined in the prover. A verification key is required to create a zkApp account.

You can also use the verifier function or verification key to verify proofs off-chain.

## Deploy a smart contract

> Smart contracts are deployed to the Mina network using the Mina zkApp CLI. The deployment process sends a transaction that contains the verification key to an address on the Mina blockchain.

Author: ```suggestion
Smart contracts are deployed to the Mina network using the Mina zkApp CLI. The deployment process sends a transaction that contains the verification key to an address on the Mina blockchain.
```
```suggestion
Smart contracts are deployed to the Mina network using the Mina zkApp CLI. The deployment process sends a transaction that contains the verification key, an output of the verifier function, to an address on the Mina blockchain.
```
